README.md

```diff
@@ -1,57 +1,364 @@
+- After signing out, click the link to `ID Token Details` to observe that the app displays a `401: unauthorized` error instead of the ID token claims when the user is not authorized.
+
+> :information_source: Did the sample not work for you as expected? Did you encounter issues trying this sample? Then please reach out to us using the [GitHub Issues](../../issues) page.
+
+## About the code
+
+This sample uses the [Microsoft Authentication Library \(MSAL\) for Python](https://github.com/AzureAD/microsoft-authentication-library-for-python) to sign in a user and obtain a token for MS Graph API. It leverages the IdentityWebPython class found in the [Microsoft Identity Python Samples Common](https://github.com/azure-samples/ms-identity-python-samples-common) repository to allow for quick app setup.
+
+In `Sample/settings.py` class:
+
+1. A configuration object is parsed from [aad.config.json](./aad.config.json)
+1. The Azure AD configuration object are used to instantiate **IdentityWebPython**
+   1. This **must** be named `MS_IDENTITY_WEB`
+1. The `MsalMiddleware` class from the ms_identity_web package is added to the project's middleware
+
+   ```python
+    from ms_identity_web.configuration import AADConfig
+    from ms_identity_web import IdentityWebPython
+    AAD_CONFIG = AADConfig.parse_json(file_path='aad.config.json')
+    MS_IDENTITY_WEB = IdentityWebPython(AAD_CONFIG)
+    ERROR_TEMPLATE = 'auth/{}.html' # for rendering 401 or other errors from msal_middleware
+    MIDDLEWARE.append('ms_identity_web.django.middleware.MsalMiddleware')
+   ```
+
+1. In the app's `Sample/urls.py` module, the necessary MSAL endpoints are hooked up:
+
+    ```python
+    msal_urls = MsalViews(settings.MS_IDENTITY_WEB).url_patterns()
+    ...
+    path(f'{settings.AAD_CONFIG.django.auth_endpoints.prefix}/', include(msal_urls)),
+    ```
+
+- The above code sets up middlwares and hooks up all necessary endpoints for the authentication process into your Django app under a route prefix (`/auth` by default). For example, the redirect endpoint is found at `/auth/redirect`.
+- When a user navigates to `/auth/sign_in` and completes a sign-in attempt, the resulting identity data is put into the session, which can be accessed through the request object at `request.identity_context_data`.
+- When an endpoint is decorated with `@ms_identity_web.login_required`, the application only allows requests to the endpoint from authenticated (signed-in) users. If the user is not signed-in, a `401: unauthorized` error is thrown, and the browser is redirected to the 401 handler.
+
+    ```python
+    @ms_identity_web.login_required
+    def a_protected_route():
+      return "if you can see this, you're signed in!"
+    ```
+
+### Calling MS Graph
+
+When a user navigates to the `/call_ms_graph` endpoint, the following code is executed:
+
+```python
+def call_ms_graph():
+    ms_identity_web.acquire_token_silently()
+    graph = 'https://graph.microsoft.com/v1.0/users'
```

Comment:
>graph [](start = 4, length = 5)

checking again, can't we use the Graph python SDK?

Reply:
GraphSDK (regular) is deprecated. The readme says to use REST calls.
[Get started with Graph - Python](https://developer.microsoft.com/en-us/graph/get-started/python) doesn't mention an SDK.


HOWEVER - Looks like there is a thing called [GraphSDKCore](https://github.com/microsoftgraph/msgraph-sdk-python-core). It looks like it is still in development.

I will create an issue in the new repo to make use of this when it is released?